scio-test/parquet/src/test/scala/com/spotify/scio/testing/ParquetTestUtilsTest.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2024 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.testing
+
+import com.spotify.scio.avro.TestRecord
+import org.apache.avro.SchemaBuilder
+import org.apache.avro.generic.GenericRecordBuilder
+import org.apache.parquet.filter2.predicate.FilterApi
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.jdk.CollectionConverters._
+
+case class SomeRecord(intField: Int)
+
+class ParquetTestUtilsTest extends AnyFlatSpec with Matchers with ParquetTestUtils {
+
+  "Avro SpecificRecords" should "be filterable and projectable" in {
+    val records = (1 to 10).map(i =>
+      new TestRecord(
+        i,
+        i.toLong,
+        i.toFloat,
+        i.toDouble,
+        true,
+        "hello",
+        List[CharSequence]("a", "b", "c").asJava
+      )
+    )
+
+    val transformed = records
+      .parquetFilter(
+        FilterApi.gt(FilterApi.intColumn("int_field"), 5.asInstanceOf[java.lang.Integer])
+      )
+      .parquetProject(
+        SchemaBuilder.record("TestRecord").fields().optionalInt("int_field").endRecord()
+      )
```

Comment:
agree it's a bit clunky! Plus, I'm now working on the TFExample integration now, and discovering that the `parquetFilter`/`parquetProject` APIs don't work for this case as Examples require an explicit Schema to be passed, so it's getting messy

Custom matchers would be nice syntactically. Something like:


```scala
records withProjection(...) should ...
records withFilter(...) should ...
```
?

Reply:
The only issues is that, IMO, many users will want to plug this directly into their JobTests, to verify that the projection won't generate a NPE in the scio job logic, for example. This would be a bit harder to do with the scalatest matcher approach

Reply:
Yeah this is the primary use-case IMO; users want to generate unfiltered data and simulate the filter/projection being applied in the pipeline

Reply:
Reviving this thread! I guess a custom matcher could work if it supported Iterables, ie:

```scala 
withPredicate(records, predicate) should  { filteredRecords => JobTest.input(ParquetAvroIO(path), filteredRecords }
```

but it is a bit awkward to use. I think the primary use case of these helpers is to simply make sure that the protection/predicate is compatible with the Scio job logic. Which actually points back to us supporting it natively in Parquet*IO in JobTest 😅😅

Reply:
I was thinking about this more... maybe it makes the most sense to implement this as a non-default Coder in scio-test-parquet, that can be constructed by the user with a projection/predicate:

```scala
def parquetTestCoder[T <: SpecificRecord](projection: Schema, predicate: Option[FilterPredicate]): Coder[T] = ...
```

That way, it could be declared in whatever scope the user needs it, either in JobTest:

```scala
implicit val parquetCoder: Coder[MyRecord] = parquetTestCoder[MyRecord](projection, Some(FilterApi.lt(...))

JobTest[MyJob.type]
  .input(ParquetAvroIO[MyRecord]("path"), records)
...
```

Or just ad-hoc:

```scala
implicit val parquetCoder: Coder[MyRecord] = parquetTestCoder[MyRecord](projection, Some(FilterApi.lt(...))

val record: MyRecord = ???

record coderShould ...
```

wdyt?

Reply:
I briefly tried out this idea and realized it won't work well as a Coder that's applied per-element: if application of `FilterPredicate` filters out the record, Coder#decode will return null. Thus, the user code would have to be written to handle null records:

```scala
implicit val parquetCoder: Coder[MyRecord] = parquetTestCoder[MyRecord](projection, Some(FilterApi.lt(...))

sc
  .parquetAvroFile[T](path)
  .filter(_ != null)
```

which is a bad pattern to enforce. so I think this idea is out.

Reply:
what about leveraging [custom scalatest matchers](https://www.scalatest.org/user_guide/using_matchers#usingCustomMatchers) with smth like

```scala
val elements: Seq[Model]
val expected: Seq[ProjectedModel]
val projection: Schema
val predicate: FilterApi

elements should withProjection(projection) and withPredicate(predicate) be expected
```
under the hood we can use the coder strategy that filters null values. WDYT?

Reply:
I think we need to update the test `ParquetIO`s to accept projection and predicate if we want to enable this testing in `JobTest`

Reply:
that would be a neat solution, since it would allow the user to opt into this functionality in JobTest. we'd have to create test `ParquetIO` classes, though, probably one for each type (Avro/Typed/Example). Or maybe a single `ParquetReadIO` class with different constructors for each type.

Reply:
I think this is getting too complicated--IMO, let not wire it into JobTest by default, but keep these as test helpers that play nicely with scalatest (I think the custom scalatest matcher route doesn't 100% work here because `withProjection`/`withPredicate` aren't strictly Matchers or Assertions). I refactored the API to work like:

```scala
records withFilter filter withProjection projection should have size(...)
records withFilter filter withProjection projection should containInAnyOrder(...)
```

It can also be used explicitly with JobTest, by just applying it to a test input.